File: joptimize/test/src/examples/narrow/MergeReturn.scala
package joptimize.examples.narrow

object MergeReturn {
  def main(x: Int, y: Int): Int = {
    if (call(x, y) == Bar) 1 else 2
  }

  def call(x: Int, y: Int): Foo = {
    if (x > y) Bar
    else Qux
  }
}

File: joptimize/test/src/examples/narrow/MergeReturn.java
package joptimize.examples.narrow;

class MergeReturn {
    @joptimize.Test(inputs = {2, 1, 1, 2})
    static int main(int x, int y) {
        return call(x, y) instanceof Bar ? 1 : 2;
    }

    static Foo call(int x, int y) {
        if (x > y) return new Bar();
        else return new Qux();
    }

    interface Foo {
        int inc(int n);
    }

    static class Bar implements Foo {
        public int inc(int n) {
            return n + 1;
        }
    }

    static class Qux implements Foo {
        public int inc(int n) {
            return n + 2;
        }
    }
}
